File: ios/QualarooMobileSdk.m
#import "QualarooMobileSdk.h"

#ifdef RCT_NEW_ARCH_ENABLED
#import "RNQualarooMobileSdkSpec.h"
#endif
@import Qualaroo;
@implementation QualarooMobileSdk
RCT_EXPORT_MODULE()

RCT_EXPORT_METHOD(init:(NSString *) apiKey) {
    [Qualaroo.shared configureWith:apiKey];
}

RCT_EXPORT_METHOD(showSurvey:(NSString *) surveyAlias) {
    [Qualaroo.shared showSurveyWith:surveyAlias on:nil forced:false delegate:nil];
}

RCT_EXPORT_METHOD(setUserId:(NSString *) userId) {
  [Qualaroo.shared setUserID:userId];
}

RCT_EXPORT_METHOD(setUserProperty:(NSString * _Nonnull)key withValue:(NSString * _Nonnull)value) {
  [Qualaroo.shared addUserProperty:key withValue:value];
}

RCT_EXPORT_METHOD(removeUserProperty:(NSString *) key) {
  [Qualaroo.shared removeUserProperty:key];
}

RCT_EXPORT_METHOD(setPreferredLanguage:(NSString *) iso2Language) {
  [Qualaroo.shared setPreferredSurveysLanguage:iso2Language error:nil];
}

// Example method
// See // https://reactnative.dev/docs/native-modules-ios
RCT_REMAP_METHOD(multiply,
                 multiplyWithA:(double)a withB:(double)b
                 withResolver:(RCTPromiseResolveBlock)resolve
                 withRejecter:(RCTPromiseRejectBlock)reject)
{
  NSNumber *result = @(a * b);

  resolve(result);
}

// Don't compile this code when we build for the old architecture.
#ifdef RCT_NEW_ARCH_ENABLED
- (std::shared_ptr<facebook::react::TurboModule>)getTurboModule:
    (const facebook::react::ObjCTurboModule::InitParams &)params
{
    return std::make_shared<facebook::react::NativeQualarooMobileSdkSpecJSI>(params);
}
#endif

@end

File: ios/QualarooMobileSdk.swift
import Qualaroo
@objc(QualarooMobileSdk)
class QualarooMobileSdk: NSObject {
    
    public static var sharedInstance = QualarooMobileSdk()
    public static var eventEmitter: QualarooEventEmiiter!
    func registerEventEmitter(eventEmitter: QualarooEventEmiiter) {
         QualarooMobileSdk.eventEmitter = eventEmitter
     }
    
    let delgate = QualarooDeligate()
    @objc(init:withResolver:withRejecter:)
    func `init`(apiKey: String, resolve:RCTPromiseResolveBlock,reject:RCTPromiseRejectBlock) -> Void {
        Qualaroo.shared.configure(with: apiKey)
    }
    
    @objc(showSurvey:withResolver:withRejecter:)
    func showSurvey(surveyAlias: String, resolve:RCTPromiseResolveBlock,reject:RCTPromiseRejectBlock) -> Void {
        Qualaroo.shared.showSurvey(with: surveyAlias,delegate:delgate )
    }

    @objc(setUserId:withResolver:withRejecter:)
    func setUserId(userId: String, resolve:RCTPromiseResolveBlock,reject:RCTPromiseRejectBlock) -> Void {
        Qualaroo.shared.setUserID(userId)
    }
    
    @objc(setUserProperty:withB:withResolver:withRejecter:)
    func setUserProperty(key: String,value: String, resolve:RCTPromiseResolveBlock,reject:RCTPromiseRejectBlock) -> Void {
        Qualaroo.shared.addUserProperty(key, withValue: value)
    }
    
    @objc(removeUserProperty:withResolver:withRejecter:)
    func removeUserProperty(key: String, resolve:RCTPromiseResolveBlock,reject:RCTPromiseRejectBlock) -> Void {
        Qualaroo.shared.removeUserProperty(key)
    }
    
    @objc(setPreferredLanguage:withResolver:withRejecter:)
    func setPreferredLanguage(iso2Language: String, resolve:RCTPromiseResolveBlock,reject:RCTPromiseRejectBlock) -> Void {
        do {
            try Qualaroo.shared.setPreferredSurveysLanguage(iso2Language)
        } catch _ {}
       
    }


  @objc(multiply:withB:withResolver:withRejecter:)
  func multiply(a: Float, b: Float, resolve:RCTPromiseResolveBlock,reject:RCTPromiseRejectBlock) -> Void {
    resolve(a*b)
  }
}

class QualarooDeligate: SurveyDelegate{

    func surveyDidStart(surveyAlias: String) {
        QualarooMobileSdk.eventEmitter.sendEvent(withName: "survey_shown", body: surveyAlias)
    }
    
    func surveyDidDismiss(surveyAlias: String) {
        QualarooMobileSdk.eventEmitter.sendEvent(withName: "survey_dismissed", body: surveyAlias)
    }
    
    func surveyDidFinish(surveyAlias: String) {
        QualarooMobileSdk.eventEmitter.sendEvent(withName: "survey_finished", body: surveyAlias)
    }

    func surveyDidClose(errorMessage: String) {
        
    }
}
